examples/src/test/scala/JedisExample.scala

```diff
@@ -58,83 +60,129 @@ object DataSources {
+      implicit S: Sync[F]): Resource[F, ObjectOutputStream] =
+    Resource.fromAutoCloseable(S.delay(new ObjectOutputStream(b)))
+
+  def inputStream[F[_]](b: ByteArrayInputStream)(
+      implicit S: Sync[F]): Resource[F, ObjectInputStream] =
+    Resource.fromAutoCloseable(S.delay(new ObjectInputStream(b)))
+
   def fromString(s: String): Array[Byte] =
     s.getBytes(Charset.forName("UTF-8"))
 
-  def serialize[F[_]: Sync, A](v: A): F[Array[Byte]] = {
-    val baos = new ByteArrayOutputStream()
-    val oos  = new ObjectOutputStream(baos)
-    Sync[F].bracket(Sync[F].pure(oos))({ in =>
-      Sync[F].pure({
-        in.writeObject(v)
-        in.flush()
-        baos.toByteArray
-      })
-    })({ in =>
-      Sync[F].delay(in.close())
+  def serialize[F[_], A](obj: A)(
+      implicit S: Sync[F]
+  ): F[ByteArray] = {
+    type Resources = (ObjectOutputStream, ByteArrayOutputStream)
+
+    def resources[F[_]](implicit S: Sync[F]): Resource[F, Resources] =
+      for {
+        byte <- byteOutputStream
+        out  <- outputStream(byte)
+      } yield (out, byte)
+
+    resources.use({
+      case (out, byte) =>
+        S.delay {
+          out.writeObject(obj)
+          out.flush()
+          byte.toByteArray
+        }
     })
   }
 
-  def deserialize[F[_]: Sync, I, A](bs: Option[Array[Byte]]): F[Option[A]] =
-    bs match {
-      case None => Sync[F].pure(None)
-      case Some(raw) => {
-        val bais = new ByteArrayInputStream(raw)
-        val oos  = new ObjectInputStream(bais)
-        Sync[F].bracket(
-          Sync[F].pure((bais, oos))
-        )({
-          case (b, o) =>
-            Sync[F].pure({
-              val res = o.readObject()
-              Option(res.asInstanceOf[A])
-            })
-        })({
-          case (b, o) =>
-            Sync[F].delay({
-              b.close()
-              o.close()
-            })
-        })
-      }
-    }
-
+  def deserialize[F[_], A](bin: ByteArray)(
+      implicit S: Sync[F]
+  ): F[Option[A]] = {
+    type Resources = (ObjectInputStream, ByteArrayInputStream)
+
+    def resources[F[_]](implicit S: Sync[F]): Resource[F, Resources] =
+      for {
+        byte <- byteInputStream(bin)
+        in   <- inputStream(byte)
+      } yield (in, byte)
+
+    resources.use({
+      case (in, byte) =>
+        S.delay {
+          val obj = in.readObject()
+          Try(obj.asInstanceOf[A]).toOption
+        }
+    })
+  }
 }
 
 case class RedisCache[F[_]: Sync](host: String) extends DataCache[F] {
-  private val binaryClient = new BinaryJedis(host)
+  private val pool = new JedisPool(host)
 
-  def client: F[BinaryJedis] =
-    Sync[F].pure(binaryClient)
+  def connection: Resource[F, Jedis] =
+    Resource.fromAutoCloseable(Sync[F].delay(pool.getResource))
 
   private def get(i: Array[Byte]): F[Option[Array[Byte]]] =
-    Sync[F].bracket(client)({ c =>
-      Sync[F].pure(Option(c.get(i)))
-    })({ in =>
-      Sync[F].pure(in.close())
+    connection.use({ c =>
```

Comment:
```suggestion
    connection.use(c => Sync[F].delay(Option(c.get(i))) )
```
You can save the braces.